mage_ai/data_preparation/models/pipeline.py

```diff
@@ -1106,6 +1167,87 @@ async def update(self, data, update_content=False):
+
+            # check if pipeline exists with same uuid and generate new one if necessary
+            if not overwrite:
+                uuid = config['uuid']
+                index = 0
+                while self.exists(uuid):
+                    index += 1
+                    uuid = f'{config["uuid"]}_{index}'
+                config['uuid'] = uuid
+                config['name'] = uuid
+            pipeline_name = config['uuid']
+
+            # retain block upstream and downstream references
+            keys_to_keep = ['upstream_blocks', 'downstream_blocks']
+            block_hierarchy = {
+                block['uuid']: {key: [] for key in keys_to_keep}
+                for block in config['blocks']
+            }
+
+            for b_index, block in enumerate(config['blocks']):
+                name = block['name']
+                uuid = block['uuid']
+                type = block['type']
+                language = block.get('language', 'python')
+
+                block_inst = Block(name=name, uuid=uuid, block_type=type, language=language)
+
+                # check if block exists with same uuid and generate new one if necessary
+                if not overwrite:
+                    index = 0
+                    while block_inst.exists:
+                        index += 1
+                        block_inst.uuid = f'{block["uuid"]}_{index}'
+
+                # save block
+                block_destination_path = block_inst.file_path
+                file_extension = block_destination_path.split('.')[-1]
+                block_directory = block_destination_path.split(os.path.sep)[-2]
+                block_zip_path = self.__find_pipeline_file(
+                    self, tmp_dir, f'{uuid}.{file_extension}', block_directory
+                )
+
+                if block_zip_path is None or not os.path.exists(block_zip_path):
+                    raise InvalidPipelineZipError(f'Block {uuid} missing from zip file.')
+
+                files_to_be_written.append((block_zip_path, block_destination_path))
+
+                # save new block parameters
+                block['uuid'] = block_inst.uuid
+                block['name'] = block_inst.uuid
+                config['blocks'][b_index] = block
+
+                # modify upstream and downstream referencesd
+                for upstr_name in block['upstream_blocks']:
+                    block_hierarchy[upstr_name]['downstream_blocks'].append(block['uuid'])
+                for downstr_name in block['downstream_blocks']:
+                    block_hierarchy[downstr_name]['upstream_blocks'].append(block['uuid'])
+
+            # save upstream and downstream references
+            hierarchy_list = list(block_hierarchy.values())
+            for b_index, block in enumerate(config['blocks']):
+                block['upstream_blocks'] = hierarchy_list[b_index]['upstream_blocks']
+                block['downstream_blocks'] = hierarchy_list[b_index]['downstream_blocks']
```

Comment:
can we just get the value from `block_hierarchy` wtih the block uuid instead of turning it into a list?

Reply:
No, because the block uuid might get updated in the config dictionary (in situations where `overwrite` is set to false and a block with that same name already exists; eg: load_titanic will become load_titanic_1). What doesn't change is the block order, so that's why I'm converting the dictionary into a list and using the index to find the appropriate block. It may be a bit overtly complicated so I am open to suggestions; I'm going to do some brainstorming on my own to see if I can simplify this a bit. One approach I see at a first glance is appending a 'original_name' parameter to each block in the config dictionary; that way I could quickly query the dictionary for the appropriate upstream/downstream information without first converting it into a list.